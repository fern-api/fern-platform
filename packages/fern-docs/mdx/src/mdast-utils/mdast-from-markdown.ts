import type { Root as MdastRoot } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { gfmFromMarkdown } from "mdast-util-gfm";
import { mathFromMarkdown } from "mdast-util-math";
import { mdxFromMarkdown } from "mdast-util-mdx";
import { gfm } from "micromark-extension-gfm";
import { math } from "micromark-extension-math";
import { mdxjs } from "micromark-extension-mdxjs";
import { UnreachableCaseError } from "ts-essentials";

export function mdastFromMarkdown(
  content: string,
  format: "mdx" | "md" = "mdx"
): MdastRoot {
  if (format === "md") {
    return fromMarkdown(content, {
      extensions: [math(), gfm()],
      mdastExtensions: [mathFromMarkdown(), gfmFromMarkdown()],
    });
  } else if (format === "mdx") {
    return fromMarkdown(content, {
      extensions: [mdxjs(), math(), gfm()],
      mdastExtensions: [
        mdxFromMarkdown(),
        mathFromMarkdown(),
        gfmFromMarkdown(),
      ],
    });
  } else {
    throw new UnreachableCaseError(format);
  }
}
