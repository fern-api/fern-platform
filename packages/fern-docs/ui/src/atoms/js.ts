import type { DocsV1Read } from "@fern-api/fdr-sdk";
import type { JsConfig } from "@fern-docs/ui";
import { isEqual } from "es-toolkit/predicate";
import { selectAtom } from "jotai/utils";
import { DOCS_ATOM } from "./docs";

export const JS_ATOM = selectAtom(DOCS_ATOM, (docs) => docs.js, isEqual);

export function withCustomJavascript(
  readShapeJsConfig: DocsV1Read.JsConfig | undefined,
  resolveFileSrc: (fileId: string) => { src: string } | undefined
): JsConfig | undefined {
  const remote = [
    ...(readShapeJsConfig?.remote ?? []),
    ...(readShapeJsConfig?.files ?? []).map((file) => ({
      url: resolveFileSrc(file.fileId)?.src,
      strategy: file.strategy,
    })),
  ].filter(isRemote);

  const toRet = {
    inline: readShapeJsConfig?.inline,
    remote: remote.length > 0 ? remote : undefined,
  };

  if (!toRet.inline && !toRet.remote) {
    return undefined;
  }

  return toRet;
}

type RemoteJs = NonNullable<JsConfig["remote"]>[number];

function isRemote(remote: {
  url: string | undefined; // potentially undefined if the file is not found
  strategy: RemoteJs["strategy"];
}): remote is RemoteJs {
  return remote.url != null;
}
