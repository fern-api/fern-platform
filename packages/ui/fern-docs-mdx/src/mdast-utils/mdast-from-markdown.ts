import type { Root as MdastRoot } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { mathFromMarkdown } from "mdast-util-math";
import { mdxFromMarkdown } from "mdast-util-mdx";
import { math } from "micromark-extension-math";
import { mdxjs } from "micromark-extension-mdxjs";
import { UnreachableCaseError } from "ts-essentials";

export function mdastFromMarkdown(content: string, format: "mdx" | "md" = "mdx"): MdastRoot {
    if (format === "md") {
        return fromMarkdown(content);
    } else if (format === "mdx") {
        return fromMarkdown(content, {
            extensions: [mdxjs(), math()],
            mdastExtensions: [mdxFromMarkdown(), mathFromMarkdown()],
        });
    } else {
        throw new UnreachableCaseError(format);
    }
}
