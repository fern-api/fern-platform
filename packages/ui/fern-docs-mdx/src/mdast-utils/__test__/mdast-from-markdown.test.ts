import { mdastFromMarkdown } from "../mdast-from-markdown.js";

describe("mdast-from-markdown", () => {
    it("should convert markdown to mdast", () => {
        const mdast = mdastFromMarkdown("Hello world");
        expect(mdast).toMatchSnapshot();
    });

    it("should convert markdown with math to mdast", () => {
        const mdast = mdastFromMarkdown("Hello world $x^2$");
        expect(mdast).toMatchSnapshot();
    });

    it("should convert markdown with flow math to mdast", () => {
        const mdast = mdastFromMarkdown("Hello world \n$$\nx^2\n$$\n");
        expect(mdast).toMatchSnapshot();
    });
});
