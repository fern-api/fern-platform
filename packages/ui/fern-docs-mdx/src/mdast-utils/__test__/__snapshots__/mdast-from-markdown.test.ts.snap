// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mdast-from-markdown > should convert markdown to mdast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "Hello world",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 1,
          "offset": 11,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdast-from-markdown > should convert markdown with flow math to mdast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "Hello world",
        },
      ],
      "position": {
        "end": {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
    {
      "data": {
        "hChildren": [
          {
            "children": [
              {
                "type": "text",
                "value": "x^2",
              },
            ],
            "properties": {
              "className": [
                "language-math",
                "math-display",
              ],
            },
            "tagName": "code",
            "type": "element",
          },
        ],
        "hName": "pre",
      },
      "meta": null,
      "position": {
        "end": {
          "column": 3,
          "line": 4,
          "offset": 22,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 13,
        },
      },
      "type": "math",
      "value": "x^2",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 5,
      "offset": 23,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdast-from-markdown > should convert markdown with math to mdast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "Hello world ",
        },
        {
          "data": {
            "hChildren": [
              {
                "type": "text",
                "value": "x^2",
              },
            ],
            "hName": "code",
            "hProperties": {
              "className": [
                "language-math",
                "math-inline",
              ],
            },
          },
          "position": {
            "end": {
              "column": 18,
              "line": 1,
              "offset": 17,
            },
            "start": {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
          },
          "type": "inlineMath",
          "value": "x^2",
        },
      ],
      "position": {
        "end": {
          "column": 18,
          "line": 1,
          "offset": 17,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 18,
      "line": 1,
      "offset": 17,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
